src/main/java/com/iota/iri/controllers/TransactionViewModel.java

```diff
@@ -379,13 +379,95 @@ public static void updateSolidTransactions(Tangle tangle, final Set<Hash> analyz
 
             transactionViewModel.updateHeights(tangle);
 
+            // trigger the update of the referenced snapshot once a tx get's solid
+            transactionViewModel.updateReferencedSnapshot(tangle);
+
             if(!transactionViewModel.isSolid()) {
                 transactionViewModel.updateSolid(true);
                 transactionViewModel.update(tangle, "solid|height");
             }
         }
     }
 
+    /**
+     * This method calculates the correct referenced snapshot and updates the transaction.
+     *
+     * It get's triggered when a transaction becomes solid and just triggers the retrieval of the referenced milestone,
+     * which then traverses the graph and tries to retrieve the values (store the values when it finds them).
+     *
+     * @param tangle database interface
+     * @throws Exception if the database update of the database fails
+     */
+    public void updateReferencedSnapshot(Tangle tangle) throws Exception {
+        // trigger the retrieval which will also cause an update
+        this.referencedSnapshot(tangle);
+    }
+
+    /**
+     * This method allows to set the referenced snapshot for this transaction.
+     *
+     * It first checks, if the value is already set and issues a database update if the value differs from the current
+     * one.
+     *
+     * @param tangle database interface
+     * @param referencedSnapshot milestone index that is referenced by this transaction
+     * @throws Exception if the database update of the database fails
+     */
+    public void referencedSnapshot(Tangle tangle, final int referencedSnapshot) throws Exception {
+        if ( referencedSnapshot != transaction.referencedSnapshot ) {
+            transaction.referencedSnapshot = referencedSnapshot;
+            update(tangle, "referencedSnapshot");
+        }
+    }
+
+    /**
+     * This method allows to retrieve the referenced snapshot of this transaction.
+     *
+     * It retrieves the minimum of the referenced snapshots (milestone indexes) of the branch and trunk transactions.
+     * This way we can easily identify transactions that have been in a parasitic side-tangle for a long time without
+     * seeing new milestones.
+     *
+     * @return snapshot index (milestone index)
+     */
+    public int referencedSnapshot(Tangle tangle) throws Exception{
+        // if the referenced snapshot is 0 -> we try to recursively determine it
+        if(transaction.referencedSnapshot == 0) {
+            // retrieve the trunk and the branch
+            TransactionViewModel trunk = this.getTrunkTransaction(tangle);
+            TransactionViewModel branch = this.getBranchTransaction(tangle);
+
+            // check if the referenced transaction is a milestone -> otherwise descend recursively
+            int trunkReferencedSnapshot;
+            if(trunk.getHash().equals(Hash.NULL_HASH)) {
```

Comment:
I usually like to either use `Objects.equals()` or put the constant in the beginning of the equality (`Hash.NULL_HASH.equals(trunk.getHash())`). This is in order to make sure no `NullPointerException`s  occur.

Unless of course you want an exception to occur in this case (which makes sense)